<!DOCTYPE html>
<html prefix="og: http://ogp.me/ns# article: http://ogp.me/ns/article#
" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>CSCI4969-6969 Assign1 | Zaki Home Page</title>
<link href="../../../assets/css/all-nocdn.css" rel="stylesheet" type="text/css">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="alternate" type="application/rss+xml" title="RSS" hreflang="en" href="../../../rss.xml">
<link rel="canonical" href="http://www.cs.rpi.edu/~zaki/courses/mlib/mlib_assign1/">
<!--[if lt IE 9]><script src="../../../assets/js/html5.js"></script><![endif]--><!-- Font Awesome --><link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
<script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"> </script><meta name="author" content="Mohammed J. Zaki">
<meta property="og:site_name" content="Zaki Home Page">
<meta property="og:title" content="CSCI4969-6969 Assign1">
<meta property="og:url" content="http://www.cs.rpi.edu/~zaki/courses/mlib/mlib_assign1/">
<meta property="og:description" content="Assign1: Transcription Factor Binding via MLP
Due Date: Feb 1, before midnight (EST 11:59:59PM)
Transcription Factors (TFs) are proteins that bind to the DNA and help
regulate gene transcription. The ">
<meta property="og:type" content="article">
<meta property="article:published_time" content="2022-01-25T13:23:01-04:00">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.10.2/dist/katex.min.css" integrity="sha384-yFRtMMDnQtDRO8rLpMIKrtPCD5jdktao2TV19YiZYWMDkUR5GQZR/NOVTdquEx1j" crossorigin="anonymous">
</head>
<body>
<a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

<!-- Menubar -->

<nav class="navbar navbar-expand-md static-top mb-4
navbar-dark
bg-dark
"><div class="container">
<!-- This keeps the margins nice -->
        <a class="navbar-brand" href="../../../">

            <span id="blog-title">Zaki Home Page</span>
        </a>
        <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false" aria-label="Toggle navigation">
            <span class="navbar-toggler-icon"></span>
        </button>

        <div class="collapse navbar-collapse" id="bs-navbar">
            <ul class="navbar-nav mr-auto">
<li class="nav-item">
<a href="../../../publications/" class="nav-link">Publications</a>
                </li>
<li class="nav-item">
<a href="https://github.com/zakimjz" class="nav-link">Github</a>
                </li>
<li class="nav-item">
<a href="../../datamining" class="nav-link">DataMining</a>
                </li>
<li class="nav-item">
<a href="../../mlib" class="nav-link">MLinBioinfo</a>

                
            </li>
</ul>
<ul class="navbar-nav navbar-right"></ul>
</div>
<!-- /.navbar-collapse -->
    </div>
<!-- /.container -->
</nav><!-- End of Menubar --><div class="container" id="content" role="main">
    <div class="body-content">
        <!--Body content-->
        
        
        
<article class="post-text storypage" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title" itemprop="headline name"><a href="." class="u-url">CSCI4969-6969 Assign1</a></h1>

        

    </header><div class="e-content entry-content" itemprop="articleBody text">
    <h2>Assign1: Transcription Factor Binding via MLP</h2>
<p><strong>Due Date</strong>: Feb 1, before midnight (EST 11:59:59PM)</p>
<p>Transcription Factors (TFs) are proteins that bind to the DNA and help
regulate gene transcription. The TFs have to recognize some "motif" on the
DNA upstream from the gene, and DNA accessibility also plays a role.</p>
<h3>MLP model</h3>
<p>In this assignment you'll modify write an MLP model the predict whether a
segments of the human chromosome 22 (Chr22) contain the binding sites for the JUND
TF.  You can modify the mlp notebook I shared with you to work on this
problem. You need to have at least one hidden layer. You have to compute a
weighted loss, and include accessibility information in your model, as
described below.</p>
<h3>Dataset</h3>
<p>The data comprises 101 length segments from Chr22, with each position a
one-hot vector denoting one of the four bases (A, C, G, T). Thus, each
element of the input is 2d with size $101 \times 4$. Each such element has
a target label $0$ or $1$, indicating whether the TF binds to that segment or not. 
The data also includes a weight per input element, since there are only a
few binding sites (0.42%), so that you'd obtain an accuracy of 99.58% just
by predicting there are no binding sites. This means you have to use the
weights to discount the losses for label $0$ and enhance the losses for
label $1$ items. 
Finally, there is an array of values, one per input element, that also
indicates the chromosome accessibility for that segment.</p>
<p>The data is split into training, validation and testing sets. Each set
contains the following files:</p>
<ul>
<li>shard-0-X.joblib: the set of 101 x 4 input elements</li>
<li>shard-0-y.joblib: the true labels: 0 or 1</li>
<li>shard-0-w.joblib: weight per input element</li>
<li>shard-0-a.joblib: accessibility value per input element</li>
</ul>
<p>You can read these files by using <strong>joblib.load</strong> function, which will
populate a numpy array. For example</p>
<pre class="code literal-block"><span></span><code><span class="n">X</span> <span class="o">=</span> <span class="n">joblib</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">'shard-0-X.joblib'</span><span class="p">)</span>
</code></pre>

<p>will results in a numpy array X, which you can then convert to torch tensor,
and so on.</p>
<p>You can download the data as <a href="https://www.cs.rpi.edu/~zaki/MLIB/data/TF_data.zip">TF_data.zip</a>. Unzip it to create the train, valid, test directories.</p>
<h3>Details</h3>
<p>First, you must write a dataset class that can load the data from the
different directories (train_dataset, valid_dataset, test_dataset).
Your class will have the following structure:</p>
<pre class="code literal-block"><span></span><code><span class="kn">from</span> <span class="nn">torch.utils.data</span> <span class="kn">import</span> <span class="n">Dataset</span>

<span class="k">class</span> <span class="nc">JUND_Dataset</span><span class="p">(</span><span class="n">Dataset</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dir</span><span class="p">):</span>
        <span class="sd">'''load X, y, w, a from data_dir'''</span>        
        <span class="nb">super</span><span class="p">(</span><span class="n">JUND_Dataset</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>

        <span class="c1"># load X, y, w, a from given data_dir</span>
        <span class="c1"># convert them into torch tensors</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">'''return len of dataset'''</span>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idx</span><span class="p">):</span>
        <span class="sd">'''return X, y, w, and a values at index idx'''</span>
</code></pre>

<p>Once you have created the dataset class it can be passed to the DataLoader
class to get batches like we did for MNIST.</p>
<p>For the MLP model, the only difference is that the 2d input ($101\times 4$)
will have to be flattened into 404d vector. Then pass it to the hidden
layer, and apply relu activation. 
You can also try to add in a dropout layer at this point.
However, before feeding the output of the
hidden layer to the output layer, you must concatenate the accessibility
value. So if you are using hidden dimension of 128, then after concatenating
the accessibility value, it will become a 129d vector, which should be fed
to the final output layer of size 1, since we have a binary class/label.</p>
<p>The only other issue is the loss function. You should use
<a href="https://pytorch.org/docs/stable/generated/torch.nn.functional.binary_cross_entropy_with_logits.html">binary_cross_entropy_with_logits</a> with weight set to the weights per
input element. Check out the documentation for details.</p>
<p>You need to train the model on the training data, and use the validation
data to select how many epochs you want to use and to choose the hidden
dimension. Use the weighted prediction accuracy as the evaluation metric.
That is, sum of the weights of the correct predictions divided by the total
weight across all the input elements. Finally, report the weighted accuracy
on the test data.</p>
<h3>What to submit</h3>
<ul>
<li>Upload your jupyter notebook on <a href="https://submitty.cs.rpi.edu/courses/s22/csci4969/gradeable/Assign1">submitty</a>
</li>
<li>The notebook must have output values for the final test accuracy.</li>
<li>Do not submit the data file or directories. </li>
</ul>
</div>
    
        
        <script src="https://cdn.jsdelivr.net/npm/katex@0.10.2/dist/katex.min.js" integrity="sha384-9Nhn55MVVN0/4OFx7EE5kpFBPsEMZxKTCnA+4fqDmg12eCTqGi6+BB2LjY8brQxJ" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/katex@0.10.2/dist/contrib/auto-render.min.js" integrity="sha384-kWPLUVMOks5AQFrykwIup5lo0m3iMkkHrD0uJ4H5cjeGihAutqP0yW0J6dpFiVkI" crossorigin="anonymous"></script><script>
                renderMathInElement(document.body,
                    {
                        
delimiters: [
    {left: "$$", right: "$$", display: true},
    {left: "\\[", right: "\\]", display: true},
    {left: "$", right: "$", display: false},
    {left: "\\(", right: "\\)", display: false}
]

                    }
                );
            </script></article><!--End of body content--><footer id="footer">
            Contents © 2022         <a href="mailto:zaki@cs.rpi.edu">Mohammed J. Zaki</a> - Powered by         <a href="https://getnikola.com" rel="nofollow">Nikola</a>         
            
            
        </footer>
</div>
</div>


        <script src="../../../assets/js/all-nocdn.js"></script><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element){var i=element.getElementsByTagName('img')[0];return i===undefined?'':i.alt;}});
    </script><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-11058802-1"></script><script>
window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());

gtag('config', 'UA-11058802-1');
</script>
</body>
</html>
